compressors/lz77.py

```diff
@@ -0,0 +1,705 @@
+        assert num_bits_consumed_encoding == encoded_block_size
+        num_bits_consumed = ENCODED_BLOCK_SIZE_HEADER_BITS + encoded_block_size
+        assert (len(combined_decoded) - 1) % 3 == 0
+        num_sequences = (len(combined_decoded) - 1) // 3
+        min_match_length = combined_decoded[0]
+        literal_counts = combined_decoded[1 : 1 + num_sequences]
+        # need to adjust lengths and offsets to undo the processing we did on the compressor
+        match_lengths = [
+            l + min_match_length
+            for l in combined_decoded[1 + num_sequences : 1 + 2 * num_sequences]
+        ]
+        match_offsets = [
+            l + 1 for l in combined_decoded[1 + 2 * num_sequences : 1 + 3 * num_sequences]
+        ]
+        lz77_sequences = [
+            LZ77Sequence(l[0], l[1], l[2])
+            for l in zip(literal_counts, match_lengths, match_offsets)
+        ]
+        return lz77_sequences, num_bits_consumed
+
+    def decode_literals(self, encoded_bitarray: BitArray):
+        """Perform entropy decoding of the literals and return the literals
+        and the number of bits consumed.
+
+        Args:
+            encoded_bitarray (BitArray): encoded bit array
+        """
+        num_bits_consumed = 0
+        # first read the size of the counts encoding
+        counts_encoding_size = bitarray_to_uint(encoded_bitarray[:ENCODED_BLOCK_SIZE_HEADER_BITS])
+        num_bits_consumed += ENCODED_BLOCK_SIZE_HEADER_BITS
+        # now decode the counts using Elias Delta
+        if counts_encoding_size == 0:
+            return [], num_bits_consumed  # no literals
+        counts, num_bits_consumed_counts = EliasDeltaUintDecoder().decode_block(
+            encoded_bitarray[num_bits_consumed : num_bits_consumed + counts_encoding_size]
+        )
+        counts = counts.data_list
+        assert counts_encoding_size == num_bits_consumed_counts
+        num_bits_consumed += counts_encoding_size
+        # generate the probability distribution from the counts
+        counts = {i: counts[i] for i in range(256) if counts[i] > 0}
+        prob_dist = ProbabilityDist.normalize_prob_dict(counts)
+
+        # load the encoding of the literals
+        literals_huffman_encoding_size = bitarray_to_uint(
+            encoded_bitarray[num_bits_consumed : num_bits_consumed + ENCODED_BLOCK_SIZE_HEADER_BITS]
+        )
+        num_bits_consumed += ENCODED_BLOCK_SIZE_HEADER_BITS
+        # decode literals with Huffman
+        decoded_literals, num_bits_consumed_literals = HuffmanDecoder(prob_dist).decode_block(
+            encoded_bitarray[num_bits_consumed : num_bits_consumed + literals_huffman_encoding_size]
+        )
+        assert literals_huffman_encoding_size == num_bits_consumed_literals
+        num_bits_consumed += literals_huffman_encoding_size
+        return decoded_literals.data_list, num_bits_consumed
+
+    def execute_lz77_sequences(self, literals: List, lz77_sequences: List[LZ77Sequence]):
```

Comment:
why is this function called `execute`.. ? 
Is this legacy, and to be consistent with other implementations?

Reply:
zstd parlance